lib/spacesuit/cors_middleware.ex

```diff
@@ -0,0 +1,166 @@
+defmodule Spacesuit.CorsMiddleware do
+  require Logger
+  use Elixometer
+
+  @http_server            Application.get_env(:spacesuit, :http_server)
+  @cors_config            Application.get_env(:spacesuit, :cors_config)
+  @supported_http_methods [:GET, :POST, :PUT, :PATCH, :DELETE, :HEAD, :OPTIONS]
+
+  @timed(key: "timed.corsMiddleware-handle", units: :millisecond)
+  def execute(req, env) do
+    case {req[:headers]["origin"], req[:method]} do
+      {nil, _} ->
+        # http://www.w3.org/TR/cors/#resource-requests
+        # § 6.1.1
+        # If the Origin header is not present terminate this set of steps.
+        Logger.debug "No origin header on request"
+        {:ok, req, env}
+
+      {origin, method} ->
+        if is_nil(origin) or !is_valid_origin(origin) do
```

Comment:
If you need more conditions in `serve_from_origin?/1` then you can use a `with` statement.